### DISCUSSION TITLE
`absl::AnyInvocable` is not guaranteed to be empty after move

### DISCUSSION THREAD

**Discussion by jiawen** [abseil/abseil-cpp#1673] [Q&A] [answered]
As the documentation in the header clearly states, when move constructing an `absl::AnyInvocable(AnyInvocable&& /*f*/)`, `f`is not guaranteed to be empty although it might be. Of course, this is perfectly valid C++, but is there any rationale for why we don't guarantee this?

It would make using the trailing `&&` variant to express "invoke-only-once" semantics less error prone to use as this issue bit me recently. Example:

```
struct Foo {
  absl::AnyInvocable<void() &&> call_once_;

  void DoThingAndNotifyAtMostOnce() {
    if (call_once_) {
      std::move(call_once_)();
      call_once_ = nullptr;  // <------ ahh! I forgot this!
    }
  }
};
```

**Comment by jiawen** [comment]
I suppose I'm actually entangling two things. I didn't dig into `absl/functional/internal/any_invocable.h` too deeply as it has a lot of deep magic. But I guess `operator(Args...)` with trailing `&&` need not be implemented using the move constructor.

**Comment by derekmauro** [comment, marked as answer]
Guaranteeing something that doesn't need to be guaranteed constrains the implementation and, for example, may prevent future optimizations.

**Reply by jiawen** [reply]
Makes sense from the perspective of "APIs are forever" and Hyrum's law.
